File: code/04-selection/python/randomized_select.py
"""RandomizedSelect: lecture-notes/code/lecture04/{java,c}의 Quickselect와 같은
반복적 3-way partition(Dutch-flag lt/scan/gt) 구조를 옮긴 것이다. 무작위인 것은
내부 pivot 선택뿐이고, 결과는 어떤 pivot을 뽑든 항상 정확한 rank번째 값이다.
rank는 0-based(0..n-1)."""
import random

# snippet:randomized-select:start
def _partition3(a, lo, hi, pivot):
    """[lt, gt) 구간이 pivot과 같다: a[lo:lt] < pivot, a[lt:gt] == pivot, a[gt:hi] > pivot."""
    lt, scan, gt = lo, lo, hi
    while scan < gt:
        if a[scan] < pivot:
            a[lt], a[scan] = a[scan], a[lt]
            lt += 1
            scan += 1
        elif a[scan] > pivot:
            gt -= 1
            a[scan], a[gt] = a[gt], a[scan]
        else:
            scan += 1
    return lt, gt


def randomized_select(a, rank):
    a = a[:]
    lo, hi = 0, len(a)
    while True:
        if hi - lo == 1:
            return a[lo]
        pivot = a[random.randrange(lo, hi)]
        lt, gt = _partition3(a, lo, hi, pivot)
        if rank < lt:
            hi = lt
        elif rank >= gt:
            lo = gt
        else:
            return pivot
# snippet:randomized-select:end

if __name__ == "__main__":
    data = [31, 8, 48, 73, 11, 3, 20, 29, 65, 15]
    print("input:", ",".join(str(x) for x in data))
    for rank in (1, 6):
        print("rank:", rank)
        print("result:", randomized_select(data, rank))

File: code/04-selection/c/randomized_select.c
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// snippet:randomized-select:start
/* lecture-notes/code/lecture04/c/quickselect.c의 quickselect_int를 그대로 옮긴
 * 것이다(RandomizedSelect). 무작위인 것은 내부 pivot 선택뿐이고, 결과는 어떤
 * pivot을 뽑든 항상 정확한 rank번째 값이다. rank는 0-based(0..n-1). */
static void swap_int(int *a, int *b) {
    int t = *a; *a = *b; *b = t;
}

/* [lo, hi)를 < pivot, == pivot, > pivot으로 나눈다. */
static void partition3(int *a, size_t lo, size_t hi, int pivot,
                       size_t *lt_out, size_t *gt_out) {
    size_t lt = lo, scan = lo, gt = hi;
    while (scan < gt) {
        if (a[scan] < pivot) swap_int(&a[lt++], &a[scan++]);
        else if (a[scan] > pivot) swap_int(&a[scan], &a[--gt]);
        else ++scan;
    }
    *lt_out = lt; *gt_out = gt;
}

static bool randomized_select(int *a, size_t n, size_t rank, int *result) {
    if (a == NULL || result == NULL || n == 0 || rank >= n) return false;
    size_t lo = 0, hi = n;
    for (;;) {
        if (hi - lo == 1) { *result = a[lo]; return true; }
        size_t pivot_index = lo + (size_t)rand() % (hi - lo);
        int pivot = a[pivot_index];
        size_t lt, gt;
        partition3(a, lo, hi, pivot, &lt, &gt);
        if (rank < lt) hi = lt;
        else if (rank >= gt) lo = gt;
        else { *result = pivot; return true; }
    }
}
// snippet:randomized-select:end

int main(void) {
    srand(20260729u);
    int data[] = {31, 8, 48, 73, 11, 3, 20, 29, 65, 15};
    int n = sizeof(data) / sizeof(data[0]);
    printf("input: ");
    for (int i = 0; i < n; i++) printf("%d%s", data[i], i + 1 < n ? "," : "\n");
    size_t ranks[] = {1, 6};
    for (int i = 0; i < 2; i++) {
        int copy[10];
        for (int j = 0; j < n; j++) copy[j] = data[j];
        int result;
        randomized_select(copy, (size_t)n, ranks[i], &result);
        printf("rank: %zu\n", ranks[i]);
        printf("result: %d\n", result);
    }
    return 0;
}
